_posts/2019-08-05-cognito-azuread-sso.markdown

```diff
@@ -0,0 +1,90 @@
+
+### An opinionated web authentication method
+
+Most large organisations invest in single sign on capabilities for their colleagues, which is typically good for security practice - provide one authentication method, and make it strong, with password rotations and multi-factor authentication.
+
+The solution we will focus on here is a pattern we commonly use when working with large enterprise clients, that have an established identity provider:
+
+- Microsoft Azure Active Directory as _Identity Provider_
+- AWS Cognito as _Authentication Service_
+- AWS Application Load Balancer as authentication proxy to our web application
+
+![cognito sso webapp design](/img/cognito-sso-design.jpg)
+
+Our example assumes a web application running on ECS or EC2 or similar, but in reality it can be anything that can update a Load Balancer Target Group.
+
+Summarising the solution:
+-  an application load balancer (**ALB**) is used in front of the web application
+- the **ALB** checks if the user is authenticated, and if not, delegates to Cognito to perform authorisation
+- **Cognito** integrates with Azure AD to achieve single sign on, and allows users to sign on as they normally would with their enterprise
+- Cognito redirects authenticated users back to the **ALB** which then passes them through to the web application
+
+### Automation: Infrastructure as Code
+
+Whilst AWS Cognito is a powerful security product, it is not without some significant shortcomings. The [inability to export or backup User Pool users](https://securityboulevard.com/2019/02/cave-of-broken-mirrors-3-issues-with-aws-cognito/) is exacerbated by the ease with which the whole user pool can be replaced (and thus, users destroyed) by simple Cloudformation changes.
+
+DevOps practitioners have also been left wanting by Cognito's poor cloudformation support. With virtually no updates since 2017, current Cloudformation support covers perhaps 50% of what CLI and SDK users can do with Cognito, leaving users to write their own custom resources for common Cognito use cases.
+
+This brings us to reason for this post - and [the supporting repository](https://github.com/MechanicalRock/cognito-sso-azure) - to provide an **infrastructure-as-code** solution for setting up and managing an AWS ALB + Cognito and Azure AD single sign on. 
+
+The repository helps you setup the following:
+- a CodePipeline for managing your cloudformation resources
+- Cognito Custom Resources to assist in (automated) configuration
+- Cognito User Pool, Pool Client, Federated SAML identity Provider, Resource Servers, etc.
+- Application Load Balancer, Listener and rules to defer authentication to Cognito
```

Comment:
inconsistent capitalisation in these dot points. `identity Provider` `Listener and rules`, etc.